graph fuser inserts explicit expands where necessary

zou3519 opened the pull request:
Fixes #10096

If the only thing preventing a simple mappable operator from being fused
into a fusion group is that its Tensor inputs are not of the same shape as the
output, then the graph fuser inserts explicit expand nodes for those
inputs.
This helps the graph fuser not miss out on any fusion opportunities
involving simple mappable operations that have Tensor inputs. This PR
doesn't do anything for the scalar case; that can be addressed later.

Test Plan
- Simple expect test case
- Added expect tests for a raw LSTMCell. The expands help speed up the
  forwards pass by allowing more operations to be fused into the LSTMCell's single
  FusionGroup.

cc @apaszke @zdevito

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -162,17 +175,33 @@ struct GraphFuser {
+  }
+
+  // Can the operator be fusable after adding adding explicit expands to inputs?
+  // XXX: Somehow de-duplicate this and isFusable.
```
This isn't maintainable until isFusible is merged with this. Already it is the case that it is being put into conditions right next to isFusible.

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -536,6 +565,46 @@ struct GraphFuser {
+    WithInsertPoint guard(node);
+
+    auto it_end = node->inputs().end();
+    if (node->matches("aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor", /*const=*/attr::alpha)) {
```
This seems fragile. Someone might add another node that has a scalar, and then forget to update this here.

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -536,6 +565,46 @@ struct GraphFuser {
+      // XXX: This hardcodes the "map size" for this FusionGroup.
+      // If we want to make the graph fuser more general in the future,
+      // we could use aten::broadcast_tensors or add a primitive op that broadcasts.
+      auto * expand = graph->insertNode(graph->create(aten::expand,
```
You can use `graph->insert` no which will make this more concise:
```
graph->insert(aten::expand, {producer, map_size, 0}, {}, ...);
```

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -565,6 +634,15 @@ struct GraphFuser {
         if(shouldFuse(consumer, producer)) {
+          // Insert explicit aten::expand nodes to inputs of simple mappable
+          // operators to make the inputs the same size, if necessary.
+          if (isSimpleMap(consumer) && !isFusable(consumer)) {
```
`shouldFuse` ... `!isFusible`: this is going to be really confusing for people to read.

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -565,6 +634,15 @@ struct GraphFuser {
+          }
+          // if producer->node() is not fusable, it must be simple mappable.
+          if (!isFusable(producer->node())) {
+            insertExplicitExpands(producer->node());
```
It might be a better idea to put the expand-producing logic into `fuse` rather than trying to make a distinction in `isFusible`. Really the first part here is just trying to establish whether we can make a mutation, and the `fuse` has traditionally done the mutate part. It would be better to keep it that way and teach `fuse` how to generate expands.

zdevito reviewed (commented):
In general, the approach makes sense but it needs some work to make the expand-generation logic less intrusive in the fusion pass. Right now it makes it increasingly difficult to make further edits to the pass. It might make sense to just establish when FusionGroups should be made and then do a followup pass to issue expands that will occur before the FusionGroup.

zou3519 committed:
graph fuser inserts explicit expands where necessary

Fixes #10096

If the only thing preventing a simple mappable operator from being fused
into a fusion group is that its Tensor inputs are not of the same shape
as the
output, then the graph fuser inserts explicit expand nodes for those
inputs.
This helps the graph fuser not miss out on any fusion opportunities
involving simple mappable operations that have Tensor inputs.

The main changes are:
- isFusable(Node) doesn't check input sizes anymore. A simple mappable
  operation with correct input/output types is fusible.
- shouldFuse(consumer, producer) checks consumer and producer for
  matching map_size.
- fuse(consumer, producer) emits expand nodes as necessary
- tryMoveChunk(consumer, chunk) also emits expand nodes as necessary because
  isFusable(Node) doesn't check input sizes anymore. A chunk should only
  be moved if the move results in a node that can be fused with
  consumer.

zou3519 committed:
New expect test; expect test updates

zou3519 commented:
I've made the following changes based on @zdevito's comments:

- isFusable(Node) doesn't check input sizes anymore. A simple mappable
  operation with correct input/output types is fusible.
- shouldFuse(consumer, producer) checks consumer and producer for
  matching map_size. If the sizes match among other checks,, then fuse(consumer, producer) is called.
- fuse(consumer, producer) emits expand nodes as necessary to fuse producer and consumer.
- tryMoveChunk(consumer, chunk) also emits expand nodes as necessary because
  isFusable(Node) doesn't check input sizes anymore. A chunk should only
  be moved if the move results in a node that can be fused with
  consumer; this is why tryMoveChunk needs to emit expand nodes.

This should be more robust than the previous approach, but I'm not sure if the expand-generation logic is still too intrusive or not.

In addition, @apaszke's upcoming refactor of the graph fuser probably will fix this exact problem, so I am also fine with punting on this PR until the refactor and seeing how the state of things are then.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -258,11 +305,58 @@ struct GraphFuser {
+    for (auto * producer: tensor_inputs) {
+      auto type = producer->type()->cast<TensorType>();
+      JIT_ASSERT(type);
+      if (std::equal(map_size.begin(), map_size.end(), type->sizes().begin())) {
```
You're not checking the length of `type->sizes()` so you can access memory past the end of this array.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -258,11 +305,58 @@ struct GraphFuser {
+      // we could use aten::broadcast_tensors or add a primitive op that broadcasts.
+      auto * expand = graph->insert(
+          aten::expand,
+          {producer, graph->insertConstant(IValue(map_size)), graph->insertConstant(0)})->node();
```
This should probably use `SymbolicVariable`.

zou3519 replied:
`SymbolicVariable` doesn't have an expand operation so I'll probably leave this as-is for now

apaszke reviewed (approved):
Minor errors that should be fixed before this lands, but generally LGTM.

zou3519 committed:
Address comments

zou3519 commented:
@pytorchbot retest this please
